File: Week 3/06. Turtle Pub & Sub/ros2_ws/src/turtle_controller/turtle_controller/turtle_controller.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from turtlesim.msg import Pose
from geometry_msgs.msg import Twist


class TurtleController(Node):
    def __init__(self):
        super().__init__("avoid_wall")
        self.publisher_ = self.create_publisher(Twist, "/turtle1/cmd_vel", 10)
        self.subscriber_ = self.create_subscription(
            Pose, "/turtle1/pose", self.pose_callback, 10)
        self.get_logger().info("Turtle has started moving..")

    def pose_callback(self, pose: Pose):
        cmd_vel = Twist()
        cmd_vel.linear.x = 2.0

        if (pose.x < 1.2 or pose.x > 9.8 or pose.y < 1.2 or pose.y > 9.5):
            cmd_vel.angular.z = 1.8
        else:
            cmd_vel.angular.z = 0.0

        self.publisher_.publish(cmd_vel)

def main(args=None):
    rclpy.init(args=args)
    avoid_wall = TurtleController()
    rclpy.spin(avoid_wall)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: Week 3 & 4/06. Turtle Pub & Sub/ros2_ws/src/pkg_cpp_py/src/turtle_controller.cpp
#include "rclcpp/rclcpp.hpp"
#include "pkg_cpp_py/turtle_controller.hpp"

AvoidWall::AvoidWall() : Node("avoid_wall")
{
    publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/turtle1/cmd_vel", 10);
    subscriber_ = this->create_subscription<turtlesim::msg::Pose>("/turtle1/pose", 10, std::bind(&AvoidWall::cmd_velocity_callback, this, _1));
}

void AvoidWall::cmd_velocity_callback(const turtlesim::msg::Pose::SharedPtr pose)
{
    cmd_vel.linear.x = 2.0;
    cmd_vel.linear.y = 0.0;

    if (pose->x < 1.2 || pose->x > 9.8 || pose->y < 1.2 || pose->y > 9.8)
    {
        RCLCPP_INFO(this->get_logger(), "Close to the wall, now turning...");
        cmd_vel.angular.z = 1.8;
    }
    else
        cmd_vel.angular.z = 0.0;

    publisher_->publish(cmd_vel);
}

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<AvoidWall>());
    rclcpp::shutdown();
    return 0;
}